### PR TITLE
Add support for BIP39 passphrase

### PR DESCRIPTION
**Pull request by beemeeupnow**
Allow [BIP39](https://en.bitcoin.it/wiki/BIP_0039#From_mnemonic_to_seed) passphrase usage.

This is needed for tallycash/extension#1022

This can be merged and released without the code in the aforementioned PR being present without breaking the extension.

This will allow loading keys with passphrases for users that choose to use them,
but will *not* store those passphrases; instead, they are accepted at deserialization
time to complete the deserialization successfully.

### CODE REVIEW

File: src/index.ts
Comment (Shadowfiend) [inline src/index.ts:98]:
```suggestion
    const passphrase = hdOptions.passphrase ?? ""
```

Let's stick to null coalescing here for clarity.

File: src/index.ts
Comment (Shadowfiend) [inline src/index.ts:103]: Above, we default `passphrase` to `""`. Prior, we passed `undefined`. Are we sure the underlying outcome in `HDNode` is identical for those two parameter values?

File: src/index.ts
Comment (Shadowfiend) [inline src/index.ts:206]: Was this not working before? If so, we should add tests that show it failing to work, as well.

**Review by Shadowfiend — CHANGES_REQUESTED** [review]
We're missing some tests here, but the code looks pretty solid. I left a couple of small remarks.

File: src/index.ts
Comment (beemeeupnow) [reply src/index.ts:103]: Yes.
https://github.com/beemeeupnow/ethers.js/blob/73d67daf9ba41d7eb71e8c7b1a0dbd8e2de4ef56/packages/hdnode/src.ts/index.ts#L314

File: src/index.ts
Comment (beemeeupnow) [reply src/index.ts:206]: No, there's an issue in ethers that will be fixed in the future.

Without this, ethers throws an error here: https://github.com/ethers-io/ethers.js/blob/73a46efea32c3f9a4833ed77896a216e3d3752a0/packages/wallet/src.ts/index.ts#L66

I have a PR open for ethers, but the maintainer recommended this change for now to continue working and I have not seen any issues from it.

Open to recommendations on tests for this one.

File: src/index.ts
Comment (beemeeupnow) [reply src/index.ts:98]: This makes perfect sense, will do.

**Comment by beemeeupnow** [comment]
Thank you for the review @Shadowfiend.

I'll look at adding some tests.

**Comment by beemeeupnow** [comment]
@Shadowfiend I added some tests. Let me know what you think.

**Comment by beemeeupnow** [comment]
@Shadowfiend,

Let me know if this looks good

Thank you

File: src/index.ts
Comment (Shadowfiend) [inline src/index.ts:28]: Feels like defaulting this to null is more in keeping with overall intent, even though it doesn't make a difference rn.

**Review by Shadowfiend — APPROVED** [review]
LGTM

File: src/index.ts
Comment (beemeeupnow) [reply src/index.ts:28]: Understandable. I made it this way so that even if the underlying library changed we would still be using the default defined by the BIP

**Comment by Shadowfiend** [comment]
Hmmm... Not sure why the main CI builds aren't running even after approval 🤔

**Review by beemeeupnow — COMMENTED** [review]
Do you have a way to manually trigger the CI?

let me re-align with main and see what happens

**Comment by beemeeupnow** [comment]
@Shadowfiend on my view of this PR page it says that I need a maintainer to approve running workflows and provides this link: https://docs.github.com/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

**Comment by Shadowfiend** [comment]
I might need to tweak a repo setting! Think I'm used to the extension repo which already has it on.

**Comment by Shadowfiend** [comment]
Ahhh no. The three builds that aren't running trigger on push instead of pull request, which never works with outside collaborators. Will fix shortly!

**Comment by Shadowfiend** [comment]
We did it! But it looks like you need to run a `yarn lint:fix` here 😬

**Review by beemeeupnow — COMMENTED** [review]
ran lint fix

File: src/index.ts
Comment (Shadowfiend) [inline src/index.ts:120]: @mhluongo forgot the one thing I wanted to ask before merging: what’s your preference on offering a secure default that doesn’t serialize pass phrase vs defaulting to serializing it?

Comment (mhluongo) [reply src/index.ts:120]: I don't use BIP-39 wallets regularly, so I'm not sure what the state of play is / what user expectations are these days. Generally I'd rather not store a passphrase if we can avoid it :thinking: 

Not doing it has some interesting implications for [`extension`](https://github.com/tallycash/extension/pull/1022) though. In particular, do BIP-39 wallets need to ask for additional passphrases over what we use for the wallet? Or is the idea that they can be the same passphrase?

Comment (Shadowfiend) [reply src/index.ts:121]: Tl;dr: After digging a bit more on passphrases, why they exist, and their intended use and security model, I have a stronger belief that hd-keyring should not store them by default. In fact, I'm tempted to say it shouldn't support storing passphrases as a first-class concept at all, instead accepting them as arguments at points where they would be necessary. I also think supporting passphrases properly in the UI will require reframing keyrings with passphrases as conceptually more similar to an external signer similar to a hardware wallet. More below.

> If we consider this place 'safe enough' to store the mnemonic, why not the data that gets combined with it to derive addresses?

Unless I'm missing something, the answer here is straightforward: if it were equally secure to store mnemonic and passphrase alongside each other, why would we have a passphrase at all? Presumably the idea is that the private key material is stored in one way, and the passphrase in a different, orthogonal way. If they're stored alongside each other the passphrase gives no additional security, right?

See [this guide from Trezor](https://blog.trezor.io/passphrase-the-ultimate-protection-for-your-accounts-3a311990925b#d30c) on what the security benefits are, and [this note from the Bitcoin wiki](https://en.bitcoin.it/wiki/Seed_phrase#Two-factor_seed_phrases) on two-factor seed phrases.

> Because the extension uses deserialize when populating keyrings in unlock(), the passphrase would need to be provided there if it was not stored here because AddAddressesSync is called. Additional refactoring would likely be required on both sides.

This feels rough, but points to an overall complication with supporting these directly in the first place. They feel like they'll have an interaction more like an external signer (where additional interactions may need to occur before derivation, signing, etc, such as plugging in a USB device for a hardware wallet, and we need to remember more in the extension). Thinking of entering the passphrase as a step similar to plugging in a hardware wallet feels like a good mental model, and reflects the second-factor nature of passphrases IMO.

This should be somewhat straightforward to support once tallycash/extension#1165 is implemented, since it'll just be another signer type.

> I don't want to have to enter these every single time that I need to use them, but want to have an option to require it in the future (so users that want the extra security can have it)

Last note: this feels implementable with some levers in the extension.

Comment (beemeeupnow) [reply src/index.ts]: @Shadowfiend,
  
Thank you for the detailed and thoughtful insight!
  
I read the RFB in the PR you linked and really like the separation of duties there. That effort will definitely help with future additions to the wallet and addresses technical debt early.
  
I pushed changes to remove the storage of the passphrase. Let me know if it looks good now

**Review by Shadowfiend — APPROVED** [review]
This looks solid to me. Thanks for the constructive back and forth and for being patient here @beemeeupnow !

**Comment by beemeeupnow** [comment]
@Shadowfiend when will the next tag and package release+upload be?

**Comment by Shadowfiend** [comment]
Should have something next week I expect!

**Comment by Shadowfiend** [comment]
Oh wait this is hd-keyring sorry 🙈 Let me check with the team on this one.

**Comment by Shadowfiend** [comment]
I expect we'll fast-track a release here with this once #12 lands.

**Comment by mhluongo** [comment]
Released @beemeeupnow!

**Pull request merged**
